src/modules/manual_control/ManualControlSelector.cpp

```diff
@@ -67,27 +71,61 @@ bool ManualControlSelector::isInputValid(const manual_control_setpoint_s &input,
 {
 	// Check for timeout
 	const bool sample_from_the_past = now >= input.timestamp_sample;
-	const bool sample_newer_than_timeout = now - input.timestamp_sample < _timeout;
+	const bool sample_newer_than_timeout = now < input.timestamp_sample + _timeout;
 
 	// Check if source matches the configuration
-	const bool source_rc_matched = (_rc_in_mode == 0) && (input.data_source == manual_control_setpoint_s::SOURCE_RC);
-	const bool source_mavlink_matched = (_rc_in_mode == 1) &&
-					    (input.data_source == manual_control_setpoint_s::SOURCE_MAVLINK_0
-					     || input.data_source == manual_control_setpoint_s::SOURCE_MAVLINK_1
-					     || input.data_source == manual_control_setpoint_s::SOURCE_MAVLINK_2
-					     || input.data_source == manual_control_setpoint_s::SOURCE_MAVLINK_3
-					     || input.data_source == manual_control_setpoint_s::SOURCE_MAVLINK_4
-					     || input.data_source == manual_control_setpoint_s::SOURCE_MAVLINK_5);
-	const bool source_any_matched = (_rc_in_mode == 2);
-	const bool source_first_matched = (_rc_in_mode == 3) &&
-					  (input.data_source == _first_valid_source
-					   || _first_valid_source == manual_control_setpoint_s::SOURCE_UNKNOWN);
-
-	return sample_from_the_past && sample_newer_than_timeout && input.valid
-	       && (source_rc_matched || source_mavlink_matched || source_any_matched || source_first_matched);
+	bool match = false;
+
+	switch (_rc_in_mode) { // COM_RC_IN_MODE
+	case 0: // RC Transmitter only
+		match = isRc(input.data_source);
+		break;
+
+	case 1: // Joystick only
+		match = isMavlink(input.data_source);
+		break;
+
+	case 2: // RC and Joystick with fallback
+		match = true;
+		break;
+
+	case 3: // RC or Joystick keep first
+		match = (input.data_source == _first_valid_source)
+			|| (_first_valid_source == manual_control_setpoint_s::SOURCE_UNKNOWN);
+		break;
+
+	case 5: // RC priority, Joystick fallback
+		match = isRc(input.data_source) || (now > _timestamp_last_rc + _timeout);
```

Comment:
this logic is a little confusing -- it's saying the input is valid if it's RC or if RC has timed out?

Reply:
The case I found from your question I fixed in ee325c4b815bfc05161c00b6ec7db5ac3291c393